[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73510
* #73486
* __->__ #73337
* #73035

This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

dzdang committed:
Implemented support for relu in quantized conv operator using cudnn

This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    @unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34439768).

dzdang committed:
Update on "Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -249,22 +261,46 @@ void raw_cudnn_convolution_forward_out(
-                                  after_add.value().data_ptr(),
-                                  };
-      uids = std::vector<int64_t>{'x', 'y', 'w', 's', 'r', 'b', 'c', 'd', 'e'};
+      if (kReluFused) {
+        data_ptrs = std::vector<void *>{reinterpret_cast<int8_t*>(input.data_ptr()), conv_output.data_ptr(),
+                                    reinterpret_cast<int8_t*>(weight.data_ptr()),
+                                    requantize_multiplier_tensor.data_ptr(),
+                                    reinterpret_cast<int8_t*>(quantized_output.data_ptr()),
+                                    broadcasted_bias.value().data_ptr(),
+                                    bias_multiplier_tensor.value().data_ptr(),
+                                    after_scales_bias.value().data_ptr(),
+                                    after_add.value().data_ptr(),
+                                    after_relu.value().data_ptr(),
+                                    };
+        uids = std::vector<int64_t>{'x', 'y', 'w', 's', 'r', 'b', 'c', 'd', 'e', 'f'};
+      } else {
+        data_ptrs = std::vector<void *>{reinterpret_cast<int8_t*>(input.data_ptr()), conv_output.data_ptr(),
+                                    reinterpret_cast<int8_t*>(weight.data_ptr()),
+                                    requantize_multiplier_tensor.data_ptr(),
+                                    reinterpret_cast<int8_t*>(quantized_output.data_ptr()),
+                                    broadcasted_bias.value().data_ptr(),
+                                    bias_multiplier_tensor.value().data_ptr(),
+                                    after_scales_bias.value().data_ptr(),
+                                    after_add.value().data_ptr(),
+                                    };
+        uids = std::vector<int64_t>{'x', 'y', 'w', 's', 'r', 'b', 'c', 'd', 'e'};
+      }
```
can we try to eliminate the duplicated code here and below?

dzdang replied:
I can, but I would have to use std::emplace_back to insert the remaining data pointers. I think it might be slightly slower because emplace_back may require a reallocation of the size exceeds the allocated space, so that's why I had it like this. Should we change it?

dzdang replied:
actually i can avoid this issue with std::vector::reserve. i'll just reserve space for 10 pointers. so i'll make this change

jerryzh168 replied:
yes, reserve makes sense

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -312,18 +348,34 @@ void raw_cudnn_convolution_forward_out(
+
   auto requant_op = cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_POINTWISE_DESCRIPTOR)
-    .setxDesc(bias.has_value() ? sum_conv_bias_op.value().getOutputTensor() : conv_op.getOutputTensor())
+    .setxDesc(bias.has_value() ? (kReluFused ? relu_op.value().getOutputTensor() : sum_conv_bias_op.value().getOutputTensor())
+                               : (kReluFused ? relu_op.value().getOutputTensor() : conv_op.getOutputTensor()))
```
these are getting complicated, maybe we can have a variable for `bias.has_value() ? sum_conv_bias_op.value().getOutputTensor() : conv_op.getOutputTensor()` and reuse it here, this is also used in L355? (we can first check has_bias() and then kReLUFused I think)

dzdang replied:
ah yes I think that will work. I thought getOutputTensor() returned a tensor, but it looks like it returns ManagedOpaqueDescriptor, which is just a shared_ptr, so storing it should be fine efficiency-wise. i'll make the change

jerryzh168 commented on test/quantization/core/test_quantized_op.py:
```diff
@@ -4142,7 +4142,7 @@ def test_qconv2d(
            # TODO: enable bias
            use_bias=st.booleans(),
            # TODO: enable relu
```
we can remove the TODO now, same for `use_bias`

jerryzh168 reviewed (approved):
looks good! has a few readability comments inline

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34439768).

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34439768).

dzdang committed:
Update on "[Quant][core][gpu] Implemented support for relu in quantized conv operator using cudnn"


This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Summary: This PR extends the conv cudnn operator to include relu. The support is implemented using optional variables.

Testing:
uncomment
```
    unittest.skip("Local only - currently the qconv2d_cudnn op is bulid "
                   "with USE_EXPERIMENTAL_CUDNN_V8_API, we can enable the test "
                   "after it is built by default")
```
above test_qconv2d_cudnn in test/quantization/core/test_quantized_op.py.
Then execute
```
python test/test_quantization.py TestQuantizedConv.test_qconv2d_cudnn
```

Differential Revision: [D34439768](https://our.internmc.facebook.com/intern/diff/D34439768)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34439768).
